# EDA

In [5]:
# library 
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import glob 
import sys
from tkinter.font import BOLD
from collections import defaultdict

### 1. Dataset Overview

In [6]:
BOLD = '\033[1m'
BLUE = '\033[94m'
GREEN = '\033[92m'
END = '\033[0m'
RED = '\033[91m'

In [7]:
folder = "/Users/sivhoung/Desktop/Year 4 intern project/Project code /dataset"
glob.glob(f"{folder}/*.csv")

['/Users/sivhoung/Desktop/Year 4 intern project/Project code /dataset/olist_sellers_dataset.csv',
 '/Users/sivhoung/Desktop/Year 4 intern project/Project code /dataset/product_category_name_translation.csv',
 '/Users/sivhoung/Desktop/Year 4 intern project/Project code /dataset/olist_orders_dataset.csv',
 '/Users/sivhoung/Desktop/Year 4 intern project/Project code /dataset/olist_order_items_dataset.csv',
 '/Users/sivhoung/Desktop/Year 4 intern project/Project code /dataset/olist_customers_dataset.csv',
 '/Users/sivhoung/Desktop/Year 4 intern project/Project code /dataset/olist_geolocation_dataset.csv',
 '/Users/sivhoung/Desktop/Year 4 intern project/Project code /dataset/olist_order_payments_dataset.csv',
 '/Users/sivhoung/Desktop/Year 4 intern project/Project code /dataset/olist_order_reviews_dataset.csv',
 '/Users/sivhoung/Desktop/Year 4 intern project/Project code /dataset/olist_products_dataset.csv']

In [8]:
files = glob.glob(f"{folder}/*.csv")
folder = "/Users/sivhoung/Desktop/Year 4 intern project/Project code /dataset"

len(files)

9

In [9]:
for i, file in enumerate(files):
    print(f"{BLUE}File {i+1}-> {file.split('/')[-1]}{END}")

File 1-> olist_sellers_dataset.csv
File 2-> product_category_name_translation.csv
File 3-> olist_orders_dataset.csv
File 4-> olist_order_items_dataset.csv
File 5-> olist_customers_dataset.csv
File 6-> olist_geolocation_dataset.csv
File 7-> olist_order_payments_dataset.csv
File 8-> olist_order_reviews_dataset.csv
File 9-> olist_products_dataset.csv


In [10]:
data = []
for file in glob.glob(f'{folder}/*.csv'):
    file_name = file.split("/")[-1]
    df = pd.read_csv(file)
    row, column = df.shape
    data.append({'File':file_name,'Rows':row,'Columns':column})

result_df = pd.DataFrame(data)
print(result_df.to_string(index=False))


                                 File    Rows  Columns
            olist_sellers_dataset.csv    3095        4
product_category_name_translation.csv      71        2
             olist_orders_dataset.csv   99441        8
        olist_order_items_dataset.csv  112650        7
          olist_customers_dataset.csv   99441        5
        olist_geolocation_dataset.csv 1000163        5
     olist_order_payments_dataset.csv  103886        5
      olist_order_reviews_dataset.csv   99224        7
           olist_products_dataset.csv   32951        9


In [11]:
for path in glob.glob(f"{folder}/*.csv"):
    df = pd.read_csv(path)
    print(f"{BLUE}======{path.split('/')[-1]}======{END}")
    print(f"{BOLD}shape:{END} {df.shape}")
    print(f"{BOLD}columns:{END} {list(df.columns)}")
    print(f"{BOLD}dtypes:{END} \n{df.dtypes}")
    print(f"{BOLD}Missing %:{END} \n{df.isna().mean().round(4) * 100}")
    print(f"{BOLD}Duplicated :{END} {df.duplicated().sum()}")
    print(f"\n")


======olist_sellers_dataset.csv======
shape: (3095, 4)
columns: ['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']
dtypes: 
seller_id                 object
seller_zip_code_prefix     int64
seller_city               object
seller_state              object
dtype: object
Missing %: 
seller_id                 0.0
seller_zip_code_prefix    0.0
seller_city               0.0
seller_state              0.0
dtype: float64
Duplicated : 0


======product_category_name_translation.csv======
shape: (71, 2)
columns: ['product_category_name', 'product_category_name_english']
dtypes: 
product_category_name            object
product_category_name_english    object
dtype: object
Missing %: 
product_category_name            0.0
product_category_name_english    0.0
dtype: float64
Duplicated : 0


======olist_orders_dataset.csv======
shape: (99441, 8)
columns: ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'or

#### 1.1. Statistical number

In [12]:
for path in glob.glob(f"{folder}/*.csv"):
    df = pd.read_csv(path)
    print(f"{BOLD}======{path.split('/')[-1]}======{END}")
    print(f"df.describe():\n{df.describe()}")
    print(f"\n")

======olist_sellers_dataset.csv======
df.describe():
       seller_zip_code_prefix
count             3095.000000
mean             32291.059451
std              32713.453830
min               1001.000000
25%               7093.500000
50%              14940.000000
75%              64552.500000
max              99730.000000


======product_category_name_translation.csv======
df.describe():
       product_category_name product_category_name_english
count                     71                            71
unique                    71                            71
top             beleza_saude                 health_beauty
freq                       1                             1


======olist_orders_dataset.csv======
df.describe():
                                order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7c

### 2. Data Structure 

In [13]:
#unique values of text columns
customer = pd.read_csv(f'{folder}/olist_customers_dataset.csv')
order = pd.read_csv(f'{folder}/olist_orders_dataset.csv')

find out the relationship between the files by checking the column names

In [14]:

file_columns={}
# collect each file's column names
for path in glob.glob(f"{folder}/*.csv"):
    df = pd.read_csv(path)
    file_columns[path.split("/")[-1]] = list(df.columns)

#print each file's column names
print(f"{BOLD}======File columns======{END}")
for file, columns in file_columns.items():
    print(f"{BOLD}{file}:{END} \n {columns}")
    print(f"--> Number of columns:{len(columns)}")

======File columns======
olist_sellers_dataset.csv: 
 ['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']
--> Number of columns:4
product_category_name_translation.csv: 
 ['product_category_name', 'product_category_name_english']
--> Number of columns:2
olist_orders_dataset.csv: 
 ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']
--> Number of columns:8
olist_order_items_dataset.csv: 
 ['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']
--> Number of columns:7
olist_customers_dataset.csv: 
 ['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']
--> Number of columns:5
olist_geolocation_dataset.csv: 
 ['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']
--> Number of columns

In [15]:
# find which columns apear in multiple files 
col_count = defaultdict(list)
for file, columns in file_columns.items():
    for col in columns:
        col_count[col].append(file)

print(f"\n{BOLD}======Columns in multiple files======{END}")
for col, files in col_count.items():
    if len(files) > 1:
        print(f"{BOLD}{col} {END} {RED}-> {END} {files}")
    


======Columns in multiple files======
seller_id  ->  ['olist_sellers_dataset.csv', 'olist_order_items_dataset.csv']
product_category_name  ->  ['product_category_name_translation.csv', 'olist_products_dataset.csv']
order_id  ->  ['olist_orders_dataset.csv', 'olist_order_items_dataset.csv', 'olist_order_payments_dataset.csv', 'olist_order_reviews_dataset.csv']
customer_id  ->  ['olist_orders_dataset.csv', 'olist_customers_dataset.csv']
product_id  ->  ['olist_order_items_dataset.csv', 'olist_products_dataset.csv']


#### Verifying PK and FK

In [16]:
seller = pd.read_csv(f'{folder}/olist_sellers_dataset.csv')
product_category_name_translation = pd.read_csv(f'{folder}/product_category_name_translation.csv')
order = pd.read_csv(f'{folder}/olist_orders_dataset.csv')
customer = pd.read_csv(f'{folder}/olist_customers_dataset.csv')
geolocation = pd.read_csv(f'{folder}/olist_geolocation_dataset.csv') 
order_payments = pd.read_csv(f'{folder}/olist_order_payments_dataset.csv')
order_reviews = pd.read_csv(f'{folder}/olist_order_reviews_dataset.csv')
product = pd.read_csv(f'{folder}/olist_products_dataset.csv')

In [17]:
#verify if the primary key is unique in each file
seller["seller_id"].is_unique
product_category_name_translation["product_category_name"].is_unique
order["order_id"].is_unique
customer["customer_id"].is_unique
product["product_id"].is_unique

True

In [18]:
geolocation["geolocation_zip_code_prefix"].is_unique
order_payments["order_id"].is_unique
order_reviews["order_id"].is_unique

False

In [19]:
order_item = pd.read_csv(f'{folder}/olist_order_items_dataset.csv')

#verify FK
order_item['seller_id'].isin(seller['seller_id']).all()


np.True_

In [20]:
product_category_name_translation['product_category_name'].isin(product['product_category_name']).all()

np.True_

In [21]:
order['customer_id'].isin(customer['customer_id']).all()

np.True_

In [22]:
order_item['product_id'].isin(product['product_id']).all()

np.True_

In [23]:
order_item['order_id'].isin(order['order_id']).all()

np.True_

In [24]:
order_payments['order_id'].isin(order['order_id']).all()

np.True_

In [25]:
order_reviews['order_id'].isin(order['order_id']).all()

np.True_

#### Check Cardinality
relationship 1 to 1, 1 to many, many to many

In [26]:
#How many child records can one parent have?
#How many customers have more than one order?

order_per_customer = order.groupby('customer_id')['order_id'].nunique()



order_per_customer.describe()




count    99441.0
mean         1.0
std          0.0
min          1.0
25%          1.0
50%          1.0
75%          1.0
max          1.0
Name: order_id, dtype: float64

In [27]:
if order_per_customer.max()>1:
    print(f"{order_per_customer.sum()} customers have more than one order.{BOLD}1 to Many relationship {END}")
else:
    print(f"All customers have only one order.{BOLD}1 to 1 relationship {END}")

All customers have only one order.1 to 1 relationship 


```Because you verified:

customers["customer_id"].is_unique → True
Every orders["customer_id"] exists in customers["customer_id"] → True
Every customer has only one order → 1 order per customer 
```

#### **Means that: The Olist business model is 1-to-1 between customers and orders.**

In [28]:
order_per_seller = order_item.groupby('seller_id')['order_id'].nunique()
items_per_products = order_item.groupby('product_id')['order_id'].nunique()
items_per_order = order_item.groupby('order_id')['order_item_id'].nunique()
payment_per_order = order_payments.groupby('order_id')['payment_sequential'].nunique()
reviews_per_order = order_reviews.groupby('order_id')['order_id'].nunique()



In [29]:
series_list = [
    ("order_per_seller", order_per_seller),
    ("items_per_products", items_per_products),
    ("items_per_order", items_per_order),
    ("payment_per_order", payment_per_order),
    ("reviews_per_order", reviews_per_order),
]

for name, series in series_list:

    print(f"\n{name}")
    print(series.describe())


order_per_seller
count    3095.000000
mean       32.313409
std       105.139763
min         1.000000
25%         2.000000
50%         6.000000
75%        21.500000
max      1854.000000
Name: order_id, dtype: float64

items_per_products
count    32951.000000
mean         3.108403
std          9.456937
min          1.000000
25%          1.000000
50%          1.000000
75%          2.000000
max        467.000000
Name: order_id, dtype: float64

items_per_order
count    98666.000000
mean         1.141731
std          0.538452
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         21.000000
Name: order_item_id, dtype: float64

payment_per_order
count    99440.000000
mean         1.044710
std          0.381166
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         29.000000
Name: payment_sequential, dtype: float64

reviews_per_order
count    98673.0
mean         1.0
std          0.0
min          1.0
25% 

In [30]:
relationships = {
    "Seller → Order Items": order_per_seller,
    "Product → Order Items": items_per_products,
    "Order → Order Items": items_per_order,
    "Order → Payments": payment_per_order,
    "Order → Reviews": reviews_per_order,
}

for relationship, series in relationships.items():

    if series.max() > 1:
        print(f"{relationship}: {BLUE}1-to-many{END}")

    else:
        print(f"{relationship}: {GREEN}1-to-1{END}")


Seller → Order Items: 1-to-many
Product → Order Items: 1-to-many
Order → Order Items: 1-to-many
Order → Payments: 1-to-many
Order → Reviews: 1-to-1


### 3. Missingess Analysis

In [31]:
    # Reload all required dataframes
table = {
    'customer': customer,
    'geolocation': geolocation,
    'order_item': order_item,
    'order_payments': order_payments,
    'order_reviews': order_reviews,
    'order': order,
    'product': product,
    'seller': seller,
    'product_category_name_translation': product_category_name_translation
}

In [32]:

#function for missing value and percentage
def missing(df,table_name):
    mis = df.isna().sum()
    mis_pct = df.isna().mean()*100

    result = pd.DataFrame({
        'table': table_name,
        'column':df.columns,
        'missing_count':mis.values,
        'missing_percentage': mis_pct.values
    })
    return result[result['missing_count']>0]


In [33]:
missing_results = []

for table_name, df in table.items():

    result = missing(df, table_name)

    missing_results.append(result)

In [34]:
missing_summary = pd.concat(missing_results, ignore_index=True)
print(missing_summary)

            table                         column  missing_count  \
0   order_reviews           review_comment_title          87656   
1   order_reviews         review_comment_message          58247   
2           order              order_approved_at            160   
3           order   order_delivered_carrier_date           1783   
4           order  order_delivered_customer_date           2965   
5         product          product_category_name            610   
6         product            product_name_lenght            610   
7         product     product_description_lenght            610   
8         product             product_photos_qty            610   
9         product               product_weight_g              2   
10        product              product_length_cm              2   
11        product              product_height_cm              2   
12        product               product_width_cm              2   

    missing_percentage  
0            88.341530  
1          

Missingness is not evenly distributed across the dataset.

#### 3.1 Order missingness pattern

In [35]:
#order missingness pattern：delivered customer date
order ['delivery_date_missing']= (order['order_delivered_customer_date'].isna())
order_missing = order['delivery_date_missing']
order_missing = order.groupby("order_status")["delivery_date_missing"].mean() * 100

In [36]:
order_status = pd.DataFrame({
    'missing_percentage': order_missing.values,
    'column': 'order_delivered_customer_date'
}, index=order_missing.index)

order_status.index.name = 'order_status'
print(order_status)

              missing_percentage                         column
order_status                                                   
approved              100.000000  order_delivered_customer_date
canceled               99.040000  order_delivered_customer_date
created               100.000000  order_delivered_customer_date
delivered               0.008292  order_delivered_customer_date
invoiced              100.000000  order_delivered_customer_date
processing            100.000000  order_delivered_customer_date
shipped               100.000000  order_delivered_customer_date
unavailable           100.000000  order_delivered_customer_date


approved, invoiced, processing, shipped, unavailable, canceled are  **Structural missing data**

In [42]:
order ['approved_date_missing'] = (
    order['order_approved_at'].isna()
)
order.groupby('order_status')["approved_date_missing"].mean()*100

order_status
approved         0.000000
canceled        22.560000
created        100.000000
delivered        0.014511
invoiced         0.000000
processing       0.000000
shipped          0.000000
unavailable      0.000000
Name: approved_date_missing, dtype: float64

In [41]:
order['carrier_date_missing']=(order['order_delivered_carrier_date'].isna())
order.groupby('order_status')['carrier_date_missing'].mean()*100

order_status
approved       100.000000
canceled        88.000000
created        100.000000
delivered        0.002073
invoiced       100.000000
processing     100.000000
shipped          0.000000
unavailable    100.000000
Name: carrier_date_missing, dtype: float64

In [44]:
approved_missing = (
    order.groupby("order_status")["approved_date_missing"].mean() * 100
)

carrier_missing = (
    order.groupby("order_status")["carrier_date_missing"].mean() * 100
)

delivery_missing = (
    order.groupby("order_status")["delivery_date_missing"].mean() * 100
)

In [55]:
order_missing_pattern = pd.DataFrame({
    "Approved date missing (%)": approved_missing,
    "Carrier date missing (%)": carrier_missing,
    "Customer delivery date missing (%)": delivery_missing
})

In [56]:
print(order_missing_pattern.round(2))

              Approved date missing (%)  Carrier date missing (%)  \
order_status                                                        
approved                           0.00                     100.0   
canceled                          22.56                      88.0   
created                          100.00                     100.0   
delivered                          0.01                       0.0   
invoiced                           0.00                     100.0   
processing                         0.00                     100.0   
shipped                            0.00                       0.0   
unavailable                        0.00                     100.0   

              Customer delivery date missing (%)  
order_status                                      
approved                                  100.00  
canceled                                   99.04  
created                                   100.00  
delivered                                   0.01  
inv

In [57]:
order_missing_pattern = order_missing_pattern.reset_index()

In [ ]:
#save CSV file for order missing pattern
order_missing_pattern.to_csv("order_missing_pattern.csv", index=False)

Missingness in the order timestamp fields is strongly associated with order_status.

In [59]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

for col in date_columns:
    order[col] = pd.to_datetime(order[col])

There are no approval timestamp before purchase timestamp

purchase ---> approve

In [139]:
# check : are there orders where approval happened before purchase?
invalid_approval = order[
    order['order_approved_at'] < order['order_purchase_timestamp']
].copy()

#approved --> carrier ?
invalid_carrier =order[
    order['order_delivered_carrier_date'] < order['order_approved_at']
].copy()

#carrier ---> customer delivery?
invalid_delivery = order[
    order["order_delivered_customer_date"]<
    order["order_delivered_carrier_date"] 
    
].copy()

print(f'invalid_approval: {invalid_approval.shape}')
print(f'invalid_carrier: {invalid_carrier.shape}')
print(f'invalid_delivery: {invalid_delivery.shape}')



invalid_approval: (0, 11)
invalid_carrier: (1359, 11)
invalid_delivery: (23, 11)


In [140]:
invalid_carrier["order_status"].value_counts()

order_status
delivered    1350
shipped         9
Name: count, dtype: int64

In [141]:
invalid_delivery["order_status"].value_counts()

order_status
delivered    23
Name: count, dtype: int64

In [142]:
invalid_carrier["approval_to_carrier"] = (
    invalid_carrier["order_delivered_carrier_date"]-
    invalid_carrier["order_approved_at"]
    
    
)


In [143]:
print(f'invalid_carrier:\n {invalid_carrier["approval_to_carrier"].describe()}')

invalid_carrier:
 count                           1359
mean     -2 days +23:14:52.846946285
std        4 days 19:18:28.055754668
min              -172 days +18:44:38
25%                -2 days +22:02:36
50%                -1 days +06:49:56
75%         -1 days +22:35:04.500000
max                -1 days +23:59:39
Name: approval_to_carrier, dtype: object


In [144]:
#find outlier
invalid_carrier.loc[
    invalid_carrier["approval_to_carrier"].idxmin(),
    [
        "order_id",
        "order_status",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date"
    ]
]

order_id                         7c48bb55e8e4f7e56d412e9653db37bc
order_status                                            delivered
order_purchase_timestamp                      2018-07-16 18:40:53
order_approved_at                             2018-07-16 18:50:22
order_delivered_carrier_date                  2018-01-26 13:35:00
order_delivered_customer_date                 2018-07-23 20:04:45
Name: 25883, dtype: object

order_delivered_carrier_date contains temporal inconsistencies. 1,359 records have a carrier timestamp earlier than the approval timestamp, including extreme cases such as an approximately 172-day discrepancy.

In [145]:
invalid_carrier["carrier_year_month"] = (
    invalid_carrier["order_delivered_carrier_date"].dt.to_period("M")
)

print(
    invalid_carrier["carrier_year_month"]
    .value_counts()
    .sort_index()
)

carrier_year_month
2017-02      1
2017-04     24
2017-05     14
2017-06      4
2017-07      1
2017-08     21
2017-09     13
2017-11      1
2017-12      5
2018-01     15
2018-02     23
2018-03      2
2018-04    393
2018-05     68
2018-06     95
2018-07    588
2018-08     91
Freq: M, Name: count, dtype: int64


The temporal inconsistencies are concentrated in certain periods rather than being uniformly distributed over time.

In [146]:
invalid_delivery["carrier_to_delivery"] = (
     invalid_delivery["order_delivered_customer_date"]-
     invalid_delivery["order_delivered_carrier_date"]
   
   
)

print(invalid_delivery["carrier_to_delivery"].describe())

count                             23
mean     -4 days +17:32:41.521739131
std        3 days 17:18:29.050924326
min               -17 days +21:41:31
25%         -6 days +15:14:27.500000
50%                -2 days +08:08:08
75%                -1 days +00:39:42
max                -1 days +23:36:42
Name: carrier_to_delivery, dtype: object


23 orders have a temporal inconsistency where the customer delivery timestamp occurs before the carrier handover timestamp.

In [149]:
order["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

#### 3.2 order review missingness pattern

In [150]:
print(order_reviews.shape)
print(order_reviews.columns)
print(order_reviews.dtypes)

(99224, 8)
Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp', 'message_missing'],
      dtype='object')
review_id                  object
order_id                   object
review_score                int64
review_comment_title       object
review_comment_message     object
review_creation_date       object
review_answer_timestamp    object
message_missing              bool
dtype: object


In [151]:
print(order_reviews["review_score"].value_counts().sort_index())

review_score
1    11424
2     3151
3     8179
4    19142
5    57328
Name: count, dtype: int64


In [156]:
order_reviews['title_missing'] = order_reviews['review_comment_title'].isna()
order_reviews['message_missing'] = order_reviews['review_comment_message'].isna()

missing_by_score = (
    order_reviews.groupby('review_score')[['title_missing', 'message_missing']]
    .mean()
    .mul(100)
    .round(2)
)

missing_by_score

,title_missing,message_missing
review_score,,
1,83.60,23.45
2,84.83,31.93
3,89.93,56.51
4,90.94,68.78
5,88.39,64.15


Review comment message missingness varies substantially across review scores, suggesting that missingness may be associated with review characteristics rather than being uniformly distributed.

#### 3.3 Product missing pattern

In [157]:
product.isna().sum().sort_values(ascending=False)

product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
product_id                      0
dtype: int64

In [ ]:
product.isna().mean().mul(100).sort_values(ascending=False).round(2)

product_category_name         1.85
product_name_lenght           1.85
product_description_lenght    1.85
product_photos_qty            1.85
product_weight_g              0.01
product_length_cm             0.01
product_height_cm             0.01
product_width_cm              0.01
product_id                    0.00
category_missing              0.00
dtype: float64

In [169]:
product[product['product_category_name'].isna()]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm,category_missing
105,a41e356c76fab66334f36de622ecbd3a,NaN,NaN,NaN,NaN,650.0,17.0,14.0,12.0,True
128,d8dee61c2034d6d075997acef1870e9b,NaN,NaN,NaN,NaN,300.0,16.0,7.0,20.0,True
145,56139431d72cd51f19eb9f7dae4d1617,NaN,NaN,NaN,NaN,200.0,20.0,20.0,20.0,True
154,46b48281eb6d663ced748f324108c733,NaN,NaN,NaN,NaN,18500.0,41.0,30.0,41.0,True
197,5fb61f482620cb672f5e586bb132eae9,NaN,NaN,NaN,NaN,300.0,35.0,7.0,12.0,True
...,...,...,...,...,...,...,...,...,...,...
32515,b0a0c5dd78e644373b199380612c350a,NaN,NaN,NaN,NaN,1800.0,30.0,20.0,70.0,True
32589,10dbe0fbaa2c505123c17fdc34a63c56,NaN,NaN,NaN,NaN,800.0,30.0,10.0,23.0,True
32616,bd2ada37b58ae94cc838b9c0569fecd8,NaN,NaN,NaN,NaN,200.0,21.0,8.0,16.0,True
32772,fa51e914046aab32764c41356b9d4ea4,NaN,NaN,NaN,NaN,1300.0,45.0,16.0,45.0,True


In [170]:
metadata_missing = (
    product['product_category_name'].isna()
    & product['product_name_lenght'].isna()
    & product['product_description_lenght'].isna()
    & product['product_photos_qty'].isna()
)

metadata_missing.sum()

np.int64(610)

In [171]:
metadata_missing.sum() == product['product_category_name'].isna().sum()

np.True_

Product metadata missingness is completely overlapping across four fields, affecting 610 products (1.85% of the product records), while the physical dimension and weight fields remain observed for these products.

In [172]:
physical_missing = (
    product['product_weight_g'].isna()
    | product['product_length_cm'].isna()
    | product['product_height_cm'].isna()
    | product['product_width_cm'].isna()
)

physical_missing.sum()

np.int64(2)

In [173]:
product[physical_missing]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm,category_missing
8578,09ff539a621711667c43eba6a3bd8466,bebes,60.0,865.0,3.0,NaN,NaN,NaN,NaN,False
18851,5eb564652db742ff8f28759cd8d2652a,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,True


### 4. delivery delay

delivery_delivered_customer_date → estimated_delivery_date → delivery_delay

In [175]:
order["delivery_delay"] = (
    order["order_delivered_customer_date"]
    - order["order_estimated_delivery_date"]
)


In [177]:
order["delivery_delay"] = order["delivery_delay"].dt.total_seconds() / (24 * 60 * 60)

In [178]:
print(order['delivery_delay'])

0        -7.107488
1        -5.355729
2       -17.245498
3       -12.980069
4        -9.238171
           ...    
99436   -10.369433
99437    -1.265324
99438    -5.524803
99439   -20.018819
99440   -17.452431
Name: delivery_delay, Length: 99441, dtype: float64


In [179]:
order["delivery_delay"].isna().sum()

np.int64(2965)

In [180]:
order["delivery_delay"].isna().mean() * 100

np.float64(2.981667521444877)

In [182]:
order["delivery_delay"].describe()

count    96476.000000
mean       -11.179120
std         10.186113
min       -146.016123
25%        -16.244384
50%        -11.948941
75%         -6.390000
max        188.975081
Name: delivery_delay, dtype: float64

In [183]:
order["delivery_delay"].skew()

np.float64(2.009198100005922)

delivery_delay is strongly right-skewed (skewness ≈ 2.01), with a mean of −11.18 days and a median of −11.95 days. The distribution also contains extreme observations ranging from −146.02 to 188.98 days.

In [184]:
Q1 = order["delivery_delay"].quantile(0.25)
Q3 = order["delivery_delay"].quantile(0.75)

IQR = Q3 - Q1

Q1, Q3, IQR

(np.float64(-16.244383680555558),
 np.float64(-6.390000000000001),
 np.float64(9.854383680555557))

In [185]:
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

lower_bound, upper_bound

(np.float64(-31.025959201388893), np.float64(8.391575520833335))

In [188]:
outliers = order[
    (order["delivery_delay"] < lower_bound) |
    (order["delivery_delay"] > upper_bound)
]

len(outliers)

4875

In [189]:
len(outliers) / order["delivery_delay"].notna().sum() * 100

np.float64(5.0530701936232845)

### 5. Select feature EDA

In [190]:
order.select_dtypes(include="number").columns

Index(['delivery_delay'], dtype='object')

In [194]:
order_item.select_dtypes(include="number").columns

Index(['order_item_id', 'price', 'freight_value'], dtype='object')

In [195]:
product.select_dtypes(include="number").columns

Index(['product_name_lenght', 'product_description_lenght',
       'product_photos_qty', 'product_weight_g', 'product_length_cm',
       'product_height_cm', 'product_width_cm'],
      dtype='object')

In [196]:
order_item[["price", "freight_value"]].describe()

,price,freight_value
count,112650.000000,112650.000000
mean,120.653739,19.990320
std,183.633928,15.806405
min,0.850000,0.000000
25%,39.900000,13.080000
50%,74.990000,16.260000
75%,134.900000,21.150000
max,6735.000000,409.680000


In [197]:
order_item[["price", "freight_value"]].skew()

price            7.923208
freight_value    5.639870
dtype: float64

Both price and freight_value exhibit strong positive skewness and contain extreme values. Therefore, imputation methods that rely on distributional assumptions may behave differently from methods that are more robust to skewed distributions.

In [198]:
product[
    [
        "product_weight_g",
        "product_length_cm",
        "product_height_cm",
        "product_width_cm"
    ]
].describe()

,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,32949.000000,32949.000000,32949.000000,32949.000000
mean,2276.472488,30.815078,16.937661,23.196728
std,4282.038731,16.914458,13.637554,12.079047
min,0.000000,7.000000,2.000000,6.000000
25%,300.000000,18.000000,8.000000,15.000000
50%,700.000000,25.000000,13.000000,20.000000
75%,1900.000000,38.000000,21.000000,30.000000
max,40425.000000,105.000000,105.000000,118.000000


In [199]:
product[
    [
        "product_weight_g",
        "product_length_cm",
        "product_height_cm",
        "product_width_cm"
    ]
].skew()

product_weight_g     3.604860
product_length_cm    1.750460
product_height_cm    2.140061
product_width_cm     1.670971
dtype: float64

The product physical attributes are positively skewed, particularly product_weight_g (skewness ≈ 3.60), and contain extreme observations. This indicates that the numerical features do not follow approximately symmetric distributions.

### 5. Missingness Mechanism Analysis

**Test 1: review_score ↔ message_missing**

In [200]:
# review score and missing message
pd.crosstab(
    order_reviews["review_score"],
    order_reviews["message_missing"]
)

message_missing,False,True
review_score,,
1,8745,2679
2,2145,1006
3,3557,4622
4,5976,13166
5,20554,36774


In [217]:
# Chi-square test of independence

from scipy.stats import chi2_contingency

table = pd.crosstab(
    order_reviews["review_score"],
    order_reviews["message_missing"]
)

chi2, p, dof, expected = chi2_contingency(table)

print(f"Chi-square: {chi2:.2f}")
print(f"p-value: {p:.2e}")
print(f"Degrees of freedom: {dof}")
print(f"Minimum expected frequency: {expected.min():.2f}")

Chi-square: 8307.00
p-value: 0.00e+00
Degrees of freedom: 4
Minimum expected frequency: 1301.28


The chi-square test showed a statistically significant association between review_score and message_missing (χ²(4) = 8307.00, p < 0.001). Therefore, we reject the null hypothesis of independence.

The expected frequencies are sufficiently large, so the chi-square test is appropriate.

**Test2: order_status ↔ order_delivered_customer_date missingness**

In [213]:
order["delivery_date_missing"] = (
    order["order_delivered_customer_date"].isna()
)

In [214]:
pd.crosstab(
    order["order_status"],
    order["delivery_date_missing"]
)

delivery_date_missing,False,True
order_status,,
approved,0,2
canceled,6,619
created,0,5
delivered,96470,8
invoiced,0,314
processing,0,301
shipped,0,1107
unavailable,0,609


In [216]:
from scipy.stats import chi2_contingency

table = pd.crosstab(
    order["order_status"],
    order["delivery_date_missing"]
)

chi2, p, dof, expected = chi2_contingency(table)

print(f"Chi-square: {chi2:.2f}")
print(f"p-value: {p:.2e}")
print(f"Degrees of freedom: {dof}")
print(f"Minimum expected frequency: {expected.min():.2f}")

Chi-square: 98959.05
p-value: 0.00e+00
Degrees of freedom: 7
Minimum expected frequency: 0.06
